engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java

```diff
@@ -9,5 +9,11 @@
 
 public interface ProcessingResultBuilder {
 
+  /**
+   * Resets the processing result build to its initial states (removes all follow up records,
+   * responses and commit tasks
+   */
+  void reset();
+
```

Comment:
Yes, the StreamProcessor is in charge of the lifecycle of the ProcessingResultBuilder. I think the contract should be that the result builder is empty when it is passed to the engine. So it wouldn't fill it with anything

Reply:
❓ Why not just let the engine create it? That way we don't need such a convention and the API simply restricts it

Reply:
That's a good question.

My line of thinking was: 
* The engine implements the `RecordProcessor` interface
* The platform calls the engine and provides implementations for all incoming and outgoing classes of the itnerface
* This way the platform can choose an implementation that is optimal for writing the result back to the stream

In this line of thinking, the engine would use what it is being offered. It would use the processing context to get a hold of the builder (implemented by platform) and then use the builder to generate the processing result (also implemented by the platfom).

At the same time, it falls a bit on its back if the engine is cheeky and just implements the `ProcessingResult` itself. Then it would avoid the nice builder and I think the platform still needs to have a way to handle it. I.e. the platform shouldn't expect that the returned `ProcessingResult` can be cast to a certain class.

So there is a conceptual contract which is not visible in the technical contract.

I will keep your proposal in mind as I work on the refactoring. I see merit in what you are suggesting.

I see a slight benefit in the other approach in terms of testing. If you can mock the processing builder that is passed in and used by the engine, you can observe more than just looking at the returned result. But there we stumble into different philosophies of testing again. 

But even if you just want to observe the output, it would still make it easier to inject the builder into the code and use a different builder in testing. E.g. in production we will probably use a process result which holds a couple of byte buffers, but for testing we could inject a builder which gives me a list of records, and then I can write tests against the test implementation of the process result.

Reply:
This actually makes sense as well 😄  Thanks for that explanation. ❤️ 

💭 If the platform is responsible for an implementation that is optimal for their needs (specific way of writing records and such), then it makes sense that they control the lifecycle. If they have a need to write data into it that cannot be cleared out, then they should deal with that in their implementation of the reset method.